### ISSUE TITLE
Optimize (NOT) EXISTS joins

### ISSUE THREAD

**Issue by jods4** [linq2db/linq2db#2815]
I love and I am sometimes still surprised how well linq2db optimizes (dynamic) queries that sometimes contain large unused portions. 👏 

It's not critical but there is an opportunity to remove unused OUTER JOINs and OUTER APPLY from [NOT] EXISTS queries.

Here's the full story:

I was doing some UPDATE based on queries in Oracle.
Sadly Oracle doesn't support FROM in UPDATE so linq2db has to duplicate the query into SET and WHERE EXISTS clauses, which is still much better than doing it yourself by hand! 😃 

This resulted in the following (simplified) SQL:
```sql
UPDATE T1
SET T1.X = (/* select copied here once for each field */)
WHERE
	EXISTS(
		SELECT *
		FROM
			T1 ext_2
				LEFT JOIN T2 c_7 ON c_7.ISO = Substr(ext_2.SRC_BQE_BIC, 6, 2)
				LEFT JOIN T2 c_8 ON c_8.ISO = Substr(ext_2.DES_BQE_BIC, 6, 2)
				LEFT JOIN T3 c_9 ON c_9.TC = ext_2.TREA_CENT AND c_9.BIC = ext_2.SRC_BQE_BIC 
		WHERE
			ext_2.NOT_HANDLED = 'X' AND ext_2.TRANS_CHANNEL IS NULL AND
			T1.EXT_ACC_PMT_SEQ = ext_2.EXT_ACC_PMT_SEQ
	)
```
This is mechanically correct, the full query was copied into the WHERE EXISTS.

But if it was written by hand I wouldn't have written a WHERE EXISTS at all and a tool like linq2db could notice the following optimization opportunities:

1. Unused OUTER JOINS have cardinality 1..N and do not change the existence of a result, so they can be completely removed.

2. Once removed, the only table left is `ext_2` which is the updated table that was repeated here just for the LEFT JOIN. With no such joins left it can be removed and the exists is FROM DUAL and can be removed too.

After optimization, the WHERE could be just:
```sql
UPDATE T1
SET X = ( /* subquery */)
WHERE T1.NOT_HANDLED = 'X' AND T1.TRANS_CHANNEL IS NULL
```
That would be very neat and it would be what I'd written manually.

Bonus chatter: a good SQL plan optimizer would totally ignore the LEFT JOIN from its plan for the same reason linq2db could remove them. I have observed Oracle 12 execution plan and it did _not_ drop the left joins but kept them in its plan anyway 😞 . 
So not only is the optimized request "nicer", it would also be more efficient.

### Environment details
linq2db version: *3.2.3*
Database Server: *Oracle 12*
Database Provider: *Managed ODP.NET*
Operating system: *Win 10*
.NET Framework: *5*

**Comment by sdanyliv** [comment]
Actually we have these optimisations, but tested them several years ago. Show me your original query and I'll point you how to  do that. I hope it will work otherwise I'll return to that functionality.

**Comment by jods4** [comment]
@sdanyliv Cool, thank!

This is the real linq2db query in its full glory with the gory details.

```cs
(from ext in db.T1
    from source in db.ISO_COUNTRY.LeftJoin(c => c.ISO == ext.SRC_BIC.Substring(4, 2))
    from destination in db.ISO_COUNTRY.LeftJoin(c => c.ISO == ext.DES_BIC.Substring(4,2))
    let sepa = source.SEPA && destination.SEPA
        ? source.ISO == destination.ISO
            ? EnumType.Sepa
            : EnumType.SepaCrossBorder
        : EnumType.Foreign
    from channel in db.ExternalChannels.LeftJoin(c => c.TreasuryCenter == ext.TREA_CENT
                                                   && c.BIC == ext.SRC_BIC
                                                   && c.Sepa == sepa)
    where ext.NOT_HANDLED == "X"
       && ext.TRANS_CHANNEL == null

    select channel)
    .Update(db.T1, x => new T1
    {
    TRANS_CHANNEL = ((TransChannel?)x.Trans_Channel) ?? TransChannel.Swift,
    IDF = x.Idf,
    UUID = DbFunctions.RandomUuid(),
    });
```

**Comment by sdanyliv** [comment]
Unfortunately this case will not work. But I have started several months ago implementation of `WeakLeftJoin` and `WeakInnerJoin`, which introduced for such behaviour. Implementation is on hold before I refactor optimiser to be more fast.
https://github.com/linq2db/linq2db/pull/2276

**Comment by jods4** [comment]
OK.

I am not sure the design `WeakLeftJoin` is the best approach. I suspect most devs would just put `LeftJoin` here and not get the benefit.

I am thinking of a slightly different approach and I'm gonna comment in more details in the PR.

**Comment by sdanyliv** [comment]
We cannot remove LEFT JOIN, if we are not sure that join is made by fields that are UniqueKey.
We have `HasUniqueKey` extension which defines such keys. But I have not tested that with not just fields.
```cs
var query = 
    from ext in db.T1.HasUniqueKey(ext => ext.SRC_BIC.Substring(4, 2))
    from source in db.ISO_COUNTRY.HasUniqueKey(c => c.ISO).LeftJoin(c => c.ISO == ext.SRC_BIC.Substring(4, 2))
    from destination in db.ISO_COUNTRY.HasUniqueKey(c => c.ISO).LeftJoin(c => c.ISO == ext.DES_BIC.Substring(4,2))
    let sepa = source.SEPA && destination.SEPA
        ? source.ISO == destination.ISO
            ? EnumType.Sepa
            : EnumType.SepaCrossBorder
        : EnumType.Foreign
    from channel in db.ExternalChannels.LeftJoin(c => c.TreasuryCenter == ext.TREA_CENT
                                                   && c.BIC == ext.SRC_BIC
                                                   && c.Sepa == sepa)
    where ext.NOT_HANDLED == "X"
       && ext.TRANS_CHANNEL == null

    select channel;
```
So, what is better `WeakLeftJoin` or defining such keys?

**Comment by jods4** [comment]
The goal of this specific issue is to notice that you _can_ remove it in the context of a [NOT] EXISTS, which is one building block of the Oracle UPDATE, so a not-so-uncommon case.

It could multiply lines but that wouldn't change the existence of a result, would it?

**Comment by sdanyliv** [comment]
I need to think about this. Remember me after release, maybe I'll find solution.

**Comment by jods4** [comment]
The gist of the reasoning is that inside an EXIST, you only care whether the subquery has a result or not.
An OUTER join never drops rows, only multiplies them.
So an OUTER join cannot turn a non-existent result into a new result, nor make an existent result disappear.
Hence if it's not used in some other expression, it can be dropped from the query.

**Comment by jods4** [comment]
Another benefit I just thought of: maybe you can merge other existing optimizations into this.
I'm thinking of #2490.

Basically, some APPLY or CROSS JOIN could deduce their cardinality automatically based on what they're joining to.
In #2490 I noted that aggregates without GROUP BY always result into exactly 1 row.

So if the optimizer flows this information into the join cardinality, it can then apply the same logic from this PR and there's no need for a special rule in #2490.

Some similar optimizations could be automatically made if a dynamic query results into an always false condition (e.g. X in empty list). Then the cardinality is 0 and this PR could remove the joins in some conditions:
- for INNER, CROSS the whole query will never have results (FALSE is inside EXISTS)
- for OUTER, it can be removed if unused because it won't multiply rows.

**Comment by MaceWindu** [comment]
That's good one. Only changes I see now:
- from some version we started to generate `UPDATE SET (x, y) = (select x, y from...)` to reduce amount of duplicate queries
- query doesn't work in v6 (including master) with `.InvalidOperationException: 'No coercion operator is defined between types 'Issue2815Table1' and 'Issue2815Table3'.'`

**Comment by jods4** [comment]
@MaceWindu yeah I later introduced `Sql.Row` so there are less duplicated subqueries in update now, but still two:
(1) for the WHERE to find rows that should be updated; and (2) to compute the new values.

The WHERE is still usually an EXIST, so would benefit from as many simplifications as possible as no value need to be computed.

This query (and others like it) is definitely in my codebase, so hopefully the regression can be fixed 😄

**Comment by MaceWindu** [comment]
Maybe you want to try and implement it? Should be an easy task (famous last words):
- pass a flag to ignore outer join cardinality to unused joins removal code (around here https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/SqlProvider/JoinsOptimizer.cs#L167C8-L167C38)
- calculate flag based on optimized select location in exists

**Comment by tb-mtg** [comment]
Any update on this?
